SPI API: frame format configuration option (TI vs Motorola)

tbursztyka opened the pull request:
Adding a configuration bit to switch between TI and Motorola frame formats.

Commit:
spi: Adding a new configuration bit for the frame format

So TI will no longer be the only frame format available, though it will
still be the default one. It will be possible to select the Motorola
frame format when relevant.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

carlescufi commented on include/drivers/spi.h:
```diff
@@ -114,6 +114,16 @@ extern "C" {
+ *
+ * 2 frame formats are exposed: TI and Motorola.
+ * By default, and because this is the one which is mostly used, the TI frame
+ * frame format will prevail.
```
Please add a link to somewhere where the formats are specified.

tbursztyka replied:
Many sources telling it's only a matter of CPOL/CPHA, but it seems to be a little bit more than that (first bit tx/rx timing). If it was only cpol/cpha it would be completely useless to expose such choice on hardware level. It's clearer in datasheets.

Just realized that I mixed the order, Motorola is the default... Will change that.

tbursztyka replied:
I'll clarify the documentation directly. No links seems to be worth

Commit:
dts/bindings: Add a dedicated frame format property on SPI device

Most of the time SPI devices use TI's frame format. But some may use
Motorola's. This is already taken care of in the SPI API and now it will
be possible to select the right format from DTS. Like:

    ...
    frame-format = <SPI_FRAME_FORMAT_MOTOROLA>;
    ...

This is only meant to be used for devices supporting both formats (so
the format is not hard-coded in the driver) and selected by hardware
configuration or else.  Which, in such case, it will need to use
DT_INST_PROP(<instance number>, frame-format) macro call to retrieve
the property value. Others can fully ignore it.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

erwango commented on include/drivers/spi.h:
```diff
@@ -117,6 +117,22 @@ extern "C" {
+ * The main difference is the behavior of the CS line. In Motorala it stays
+ * active the whole transfer. In TI, it's active only one serial clock period
+ * prior to actually make the transfer, it is thus inactive during the transfer
+ * which one ends when the clocks ends as well.
```
```suggestion
 * prior to actually make the transfer, it is thus inactive during the transfer,
 * which ends when the clocks ends as well.
```

erwango commented on include/drivers/spi.h:
```diff
@@ -117,6 +117,22 @@ extern "C" {
+ * prior to actually make the transfer, it is thus inactive during the transfer
+ * which one ends when the clocks ends as well.
+ * By default, and because this is the one which is mostly used, the Motorola
+ * frame format will prevail.
```
```suggestion
 * By default, as it is the most commonly used, the Motorola
 * frame format will prevail.
```

erwango commented on include/drivers/spi.h:
```diff
@@ -248,7 +264,7 @@ struct spi_cs_control {
  *     lock_on             [ 13 ]      - Keep resource locked for the caller.
  *     cs_active_high      [ 14 ]      - Active high CS logic.
- *     reserved            [ 15 ]      - reserved for future use.
+ *     format              [ 15 ]      - Motorola or TI (if set) frame format.
```
```suggestion
 *     format              [ 15 ]      - Motorola or TI frame format (optional).
```

erwango commented on dts/bindings/spi/spi-device.yaml:
```diff
@@ -29,3 +29,16 @@ properties:
+    frame-format:
+      type: int
+      required: false
+      description: |
```
You can add a `default: 0`

FRASTM commented:
@tbursztyka, could you please have a look at my PR https://github.com/zephyrproject-rtos/zephyr/pull/40860
I wonder if the structure `spi_config` is getting the frame-format correctly. I add a commit 538b0992d02b2b41ceca2c065b741e64a9594289 but It think the bit[15] of the operation is well placed for this: 
```
/**
 * @name SPI frame format
 * @{
 */
#define SPI_FRAME_FORMAT_SHIFT	(15U)
#define SPI_FRAME_FORMAT_MASK	(0x1U << SPI_FRAME_FORMAT_SHIFT)
#define SPI_FRAME_FORMAT_GET(_operation_)					\
	(((_operation_) & SPI_FRAME_FORMAT_MASK) >> SPI_FRAME_FORMAT_SHIFT)

```
